src/piadm.sh

```diff
@@ -655,13 +675,250 @@ ispoolenabled() {
+	ln -s platform-ipxe platform
+	ln -s boot-ipxe boot
+
+	# Install a PI for backup booting purposes.
+	if ! install_pi_CN "$activestamp" && \
+		[[ "$CNAPI_DEFAULT_PI" != "$activestamp" ]]; then
+		/bin/rm -rf platform-"$activestamp"
+		if ! install_pi_CN "$CNAPI_DEFAULT_PI"; then
+			/bin/rm -rf platform-"$CNAPI_DEFAULT_PI"
+			err "No PIs available"
+		fi
+	fi
+	# installstamp will be set by the successful install_pi_CN()
+
+	# Populate loader.conf.
+	# NOTE: One could uncomment the sed below and replace the cp if one
+	# wished to have the CN backup-boot not go into the Triton HN
+	# installer but act in a different kind of weird way.
+	# sed 's/headnode="true"/headnode="false"/g' \ <
+	#	boot-ipxe/loader.conf.tmpl > boot-ipxe/loader.conf
+	cp boot-ipxe/loader.conf.tmpl boot-ipxe/loader.conf
+	{
+		echo 'ipxe="true"'
+		echo 'smt_enabled="true"'
+		echo 'console="ttyb,ttya,ttyc,ttyd,text"'
+		echo 'os_console="ttyb"'
+		echo 'fstype="ufs"'
+		# use $installstamp to help!
+		echo "platform-version=$installstamp"
+		# Need an extra "platform" in here to satisfy the illumos
+		# load-time that looks for a unix in the boot archive.  The
+		# boot file path MUST have /platform/i86pc/kernel/amd64/unix
+		# as its trailing components.  See install_pi_CN() for the
+		# insertion of a symlink to help out.
+		echo bootfile=\"/platform-"$installstamp"/platform/i86pc/kernel/amd64/unix\"
+		echo boot_archive_name=\"/platform-"$installstamp"/i86pc/amd64/boot_archive\"
+		echo boot_archive.hash_name=\"/platform-"$installstamp"/i86pc/amd64/boot_archive.hash\"
+	} >> boot-ipxe/loader.conf
+
+	# Caller will invoke update_boot_sectors.
+}
+
+update_CN() {
+	declare -a pdirs
+
+	if [[ "$TRITON_CN" != "yes" ]]; then
+		err "The update command may only be used on a Triton Compute Node"
+	fi
+
+	piname_present_get_bootfs "ipxe" "$1"
+	cd "/${bootfs}" || fatal "Could not chdir to /$bootfs"
+
+	# First check if the backup PI is in need of update.
+	# The standard iPXE/CN deployment has exactly one platform-STAMP.
+	mapfile -t pdirs < <(ls -d platform-[0-9]*T*Z)
+	if [[ ${#pdirs[@]} -gt 1 ]]; then
+		corrupt "Multiple platform-STAMP in CN bootfs /${bootfs}/."
+	elif [[ ${#pdirs[@]} -lt 1 ]]; then
+		corrupt "No platform-STAMP in CN bootfs /${bootfs}/."
+	fi
+
+	pdir=${pdirs[0]}
+
+	pstamp=$(cat "${pdir}"/etc/version/platform)
+	if [[ "$pstamp" != "$activestamp" ]]; then
+		vecho "Updating backup PI to" "$activestamp"
+		if ! install_pi_CN "$activestamp" && \
+			[[ "$CNAPI_DEFAULT_PI" != "$activestamp" ]]; then
+			vecho "...trying" "$CNAPI_DEFAULT_PI" "instead"
+			if ! install_pi_CN "$CNAPI_DEFAULT_PI" ; then
+				/bin/rm -rf platform-"$activestamp"
+				/bin/rm -rf platform-"$CNAPI_DEFAULT_PI"
+				err "No PIs available, keeping" "$pstamp"
+			fi
+		fi
+
+		# Success, don't keep the old one!
+		# $installstamp will have new PI stamp, will inform below.
+		vecho "...success installing $installstamp"
+		/bin/rm -rf "$pdir"
+		tfile=$(mktemp)
```

Comment:
Maybe we should be defensive about mktemp failing? Not sure if we are running with `set -e`
`tfile=$(mktemp) || fatal "failed to create temp file"`

Reply:
I guess if mktemp fails we are probably in more trouble.  Up to you if you want to be defensive here.
If not I can mark as resolved and +1

Reply:
Honestly, if mktemp fails we have much bigger problems.